libs/core/gasnet_base/src/gasnet_environment.cpp

```diff
@@ -0,0 +1,701 @@
+    hpx::mutex gasnet_environment::dshm_mut{};
+    hpx::mutex gasnet_environment::mtx_{};
+    bool gasnet_environment::enabled_ = false;
+    bool gasnet_environment::has_called_init_ = false;
+    int gasnet_environment::provided_threading_flag_ = GASNET_PAR;
+    int gasnet_environment::is_initialized_ = -1;
+    int gasnet_environment::init_val_ = GASNET_ERR_RESOURCE;
+    hpx::mutex* gasnet_environment::segment_mutex = nullptr;
+    gasnet_seginfo_t* gasnet_environment::segments = nullptr;
+
+    ///////////////////////////////////////////////////////////////////////////
+    int gasnet_environment::init(int* argc, char*** argv, const int minimal,
+        const int required, int& provided)
+    {
+        if (!has_called_init_)
+        {
+            gasnet_environment::init_val_ = gasnet_init(argc, argv);
+            has_called_init_ = true;
+        }
+
+        if (gasnet_environment::init_val_ == GASNET_ERR_NOT_INIT)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET initialization error");
+        }
+        else if (gasnet_environment::init_val_ == GASNET_ERR_RESOURCE)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init", "GASNET resource error");
+        }
+        else if (gasnet_environment::init_val_ == GASNET_ERR_BAD_ARG)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET bad argument error");
+        }
+        else if (gasnet_environment::init_val_ == GASNET_ERR_NOT_READY)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET not ready error");
+        }
+
+        if (provided < minimal)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET doesn't provide minimal requested thread level");
+        }
+
+        if (gasnet_attach(nullptr, 0, gasnet_getMaxLocalSegmentSize(), 0) !=
+            GASNET_OK)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET failed to attach to memory");
+        }
+
+        // create a number of segments equal to the number of hardware
+        // threads per machine (locality)
+        //
+        gasnet_environment::segments = new gasnet_seginfo_t[size()];
+        gasnet_environment::segment_mutex = new hpx::mutex[size()];
+
+        GASNET_Safe(gasnet_getSegmentInfo(segments, size()));
+
+        int retval;
+
+        gasnet_barrier_notify(0, GASNET_BARRIERFLAG_ANONYMOUS);
+        while ((retval = gasnet_barrier_try(0, GASNET_BARRIERFLAG_ANONYMOUS)) ==
+            GASNET_ERR_NOT_READY)
+        {
+        }
+        gasnet_barrier_try(0, GASNET_BARRIERFLAG_ANONYMOUS);
```

Comment:
The construct above is incorrect. 

Once `gasnet_barrier_try()` returns success (last iteration of the loop), calling it again before the next `gasnet_barrier_notify()` is erroneous. The final `gasnet_barrier_try()` call should be deleted, or better yet replace all the lines after `gasnet_barrier_notify()` with `gasnet_barrier_wait(0, GASNET_BARRIERFLAG_ANONYMOUS);`






```suggestion
        gasnet_barrier_notify(0, GASNET_BARRIERFLAG_ANONYMOUS);
        gasnet_barrier_wait(0, GASNET_BARRIERFLAG_ANONYMOUS);
```

Reply:
Thank you for the correction.